Source/Core/Core/AchievementManager.cpp

```diff
@@ -152,66 +158,218 @@ void AchievementManager::LoadGameByFilenameAsync(const std::string& iso_path,
           },
       .close = [](void* file_handle) { delete reinterpret_cast<FilereaderState*>(file_handle); }};
   rc_hash_init_custom_filereader(&volume_reader);
-  if (!rc_hash_generate_from_file(m_game_hash.data(), RC_CONSOLE_GAMECUBE, iso_path.c_str()))
+  m_queue.EmplaceItem([this, callback, file_path] {
+    Hash new_hash;
+    if (!rc_hash_generate_from_file(new_hash.data(), RC_CONSOLE_GAMECUBE, file_path.c_str()))
```

Comment:
This is a race condition... Because the custom file reader in rcheevos is global state, you cannot guarantee that your filereader is still the correct one by the time this line is executed. Move the `rc_hash_filereader volume_reader` as well as the `rc_hash_init_custom_filereader(&volume_reader);` into the queue function and the lock around it (a separate, global lock from the `m_lock` should be fine, just make sure you're not holding both at the same time).

Same goes for the other function where the rcheevos filereader is used.